bee-node/src/fullnode/builder.rs

```diff
@@ -0,0 +1,438 @@
+        self.deps.insert(TypeId::of::<W>(), W::dependencies());
+        self.worker_starts.insert(
+            TypeId::of::<W>(),
+            Box::new(|node| {
+                Box::pin(async move {
+                    log::debug!("Starting worker {}...", type_name::<W>());
+                    match W::start(node, config).await {
+                        Ok(w) => node.core.add_worker(w),
+                        Err(e) => panic!("Worker `{}` failed to start: {:?}.", type_name::<W>(), e),
+                    }
+                })
+            }),
+        );
+        self.worker_stops.insert(
+            TypeId::of::<W>(),
+            Box::new(|node| {
+                Box::pin(async move {
+                    match node.core.remove_worker::<W>().stop(node).await {
+                        Ok(()) => {}
+                        Err(e) => panic!("Worker `{}` failed to stop: {:?}.", type_name::<W>(), e),
+                    }
+                })
+            }),
+        );
+        self.worker_names.insert(TypeId::of::<W>(), type_name::<W>());
+        self
+    }
+
+    /// Adds a resource to the full node.
+    fn with_resource<R: Any + Send + Sync>(mut self, res: R) -> Self {
+        self.resource_registers.push(Box::new(move |node| {
+            node.register_resource(res);
+        }));
+        self
+    }
+
+    /// Returns the built full node.
+    async fn finish(self) -> Result<FullNode<S>, Self::Error> {
+        let builder = self;
+
+        let network_spec = builder.config().network_spec();
+
+        // Print the network info the node is trying to connect to.
+        log::info!(
+            "Joining network \"{}\"({}). Bech32 hrp \"{}\".",
+            network_spec.name(),
+            network_spec.id(),
+            network_spec.hrp()
+        );
+
+        // Print the local entity data.
+        log::info!("{}", builder.config().local());
+
+        // Add the resources that are shared throughout the node.
+        let builder = add_node_resources(builder)?;
+
+        // Initialize the gossip layer.
+        let (gossip_rx, builder) = initialize_gossip_layer(builder).await?;
+
+        // Initialize autopeering (if enabled).
+        let (autopeering_rx, builder) = initialize_autopeering(builder).await?;
+
+        // Initialize the ledger.
+        let builder = initialize_ledger(builder);
+
+        // Initialize the protocol.
+        let builder = initialize_protocol(builder, gossip_rx, autopeering_rx);
+
+        // Initialize the node API.
+        let builder = initialize_api(builder).await;
+
+        // Initialize the Tangle.
+        let builder = initialize_tangle(builder);
+
+        // Start the version checker.
+        let builder = builder.with_worker::<VersionChecker>();
+
+        // Start the MQTT broker.
+        let mqtt_cfg = builder.config().mqtt_config.clone();
+        let builder = builder.with_worker_cfg::<Mqtt>(mqtt_cfg);
```

Comment:
It's great to have comments, but here they seem maybe a little bit redundant.

Reply:
Even better! 💪

Reply:
Removed excessive comments.